core/src/apps/cardano/helpers/address_credential_policy.py

```diff
@@ -0,0 +1,195 @@
+            address_parameters.address_n_staking
+        )
+
+        if not _do_base_address_credentials_match(
+            keychain,
+            address_parameters.address_n,
+            address_parameters.address_n_staking,
+            address_parameters.staking_key_hash,
+        ):
+            return (
+                CREDENTIAL_POLICY_FAIL_OR_WARN_UNUSUAL | CREDENTIAL_POLICY_WARN_MISMATCH
+                if is_unusual
+                else CREDENTIAL_POLICY_WARN_MISMATCH
+            )
+
+        return (
+            CREDENTIAL_POLICY_FAIL_OR_WARN_UNUSUAL
+            if is_unusual
+            else CREDENTIAL_POLICY_SHOW
+        )
+    elif address_type == CardanoAddressType.BASE_SCRIPT_KEY:
+        is_unusual = address_parameters.address_n_staking and not SCHEMA_STAKING.match(
+            address_parameters.address_n_staking
+        )
+        return (
+            CREDENTIAL_POLICY_FAIL_OR_WARN_UNUSUAL | CREDENTIAL_POLICY_WARN_MISMATCH
+            if is_unusual
+            else CREDENTIAL_POLICY_WARN_MISMATCH
+        )
+    elif address_type in (
+        CardanoAddressType.POINTER,
+        CardanoAddressType.POINTER_SCRIPT,
+    ):
+        return CREDENTIAL_POLICY_WARN_POINTER
+    elif address_type == CardanoAddressType.REWARD:
+        return CREDENTIAL_POLICY_SHOW_KEY_PATH
+    elif address_type in (
+        CardanoAddressType.BASE_KEY_SCRIPT,
+        CardanoAddressType.BASE_SCRIPT_SCRIPT,
+        CardanoAddressType.REWARD_SCRIPT,
+    ):
+        return CREDENTIAL_POLICY_WARN_SCRIPT
+    elif address_type in (
+        CardanoAddressType.ENTERPRISE,
+        CardanoAddressType.ENTERPRISE_SCRIPT,
+        CardanoAddressType.BYRON,
+    ):
+        return CREDENTIAL_POLICY_WARN_NO_STAKING
+    else:
+        raise ValueError("Invalid address type")
+
+
+def _do_base_address_credentials_match(
+    keychain: Keychain,
+    address_n: list[int],
+    address_n_staking: list[int],
+    staking_key_hash: bytes | None,
+) -> bool:
+    assert address_n_staking or staking_key_hash is not None
+
+    spending_account_staking_path = _path_to_staking_path(address_n)
+    if address_n_staking:
+        if address_n_staking != spending_account_staking_path:
+            return False
+    else:
+        spending_account_staking_key_hash = get_public_key_hash(
+            keychain, spending_account_staking_path
+        )
+        if staking_key_hash != spending_account_staking_key_hash:
+            return False
+    return True
+
+
+def _path_to_staking_path(path: list[int]) -> list[int]:
+    return to_account_path(path) + [CHAIN_STAKING_KEY, 0]
+
+
+def _policy_to_change_output_policy(
+    credential_policy: int,
+    signing_mode: CardanoTxSigningMode,
+) -> int:
+    """
+    Updates the policy to a tx change output policy. As a result, different text will be displayed to the user.
+    """
+    credential_policy |= CREDENTIAL_POLICY_SHOW_OUTPUT
+    credential_policy |= CREDENTIAL_POLICY_SHOW_CHANGE
```

Comment:
I'm not sure, but isn't this the only place where either of these two policies can be assigned? Can't they be merged into one? (But that might not be what was the intended behavior.)

Reply:
Hm, I think it's ok to merge them. A credential policy is (and always will be) only determined for change outputs actually. https://github.com/vacuumlabs/trezor-firmware/pull/66/commits/4d77f4e42747348ce6de09ab62a8e5ae49c07e53#diff-7774760cef1ad4636fcc2daf980fad9f3eb54dd16ef6cadeba563b245169067aR196